arm9/src/font.c

```diff
@@ -0,0 +1,271 @@
+#include "global.h"
+#include "font.h"
+#include "gf_gfx_loader.h"
+#include "render_text.h"
+#include "string16.h"
+#include "text.h"
+
+struct UnkStruct_02002C14 *UNK_02106FC8;
+
+struct UnkStruct_02002C14 UNK_02106FCC;
+
+const u16 UNK_020ECB54[] = { 0, 0, 1, 0, 2, 0, 3, 0 };
+
+const struct FontInfo gFontInfos[5] = {
+    { 0x0B, 0x10, 0x00, 0x00, 0x00, 0x01, 0x0F, 0x02 },
+    { 0x0B, 0x10, 0x00, 0x00, 0x00, 0x01, 0x0F, 0x02 },
+    { 0x0B, 0x10, 0x00, 0x00, 0x00, 0x01, 0x0F, 0x02 },
+    { 0x0B, 0x10, 0x00, 0x00, 0x00, 0x01, 0x0F, 0x02 },
+    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 },
+};
+
+extern u32 FUN_02021590(u32, u16, u32, u16, u32);
+extern void FUN_020215E0(u32, u32, u32);
+extern void FUN_020215C8(u32 param0);
+extern void FUN_02021750(void *, u32);
+extern u32 FUN_020218D8(void *, u16 *str, u32);
+
+THUMB_FUNC void FUN_02002C14()
+{
+    UNK_02106FC8 = &UNK_02106FCC;
+
+    for (u32 i = 0; i < 4; i++)
+    {
+        UNK_02106FC8->unk84[i] = NULL;
+        UNK_02106FC8->unk94[i] = NULL;
+    }
+
+    SetFontsPointer(&gFontInfos[0]);
+}
+
+THUMB_FUNC void FUN_02002C50(u32 param0, u32 param1)
+{
+    UNK_02106FC8->unk94[param0] =
+        FUN_02021590(14, UNK_020ECB54[param0 * 2], 1, UNK_020ECB54[param0 * 2 + 1], param1);
+}
+
+THUMB_FUNC void FUN_02002C84(s32 param0, u32 param1)
+{
+    GF_ASSERT(param0 < 4);
+    GF_ASSERT(UNK_02106FC8->unk94[param0] != NULL);
+
+    FUN_020215E0(UNK_02106FC8->unk94[param0], 0, param1);
+}
+
+THUMB_FUNC void FUN_02002CC0(s32 param0)
+{
+    GF_ASSERT(param0 < 4);
+    GF_ASSERT(UNK_02106FC8->unk94[param0] != NULL);
+
+    FUN_020215E0(UNK_02106FC8->unk94[param0], 1, 0);
+}
+
+#ifdef NONMATCHING
+THUMB_FUNC void FUN_02002CF8(s32 param0)
```

Comment:
The following appears to match. Note: parameter type is important, mwccarm treats `s32` and `int` differently sometimes.

```c
void FUN_02002CF8(int a0)
{
    s32 r1;

    GF_ASSERT(a0 < 4);

    if (UNK_02106FC8->unk84[a0] != NULL)
    {
        for (r1 = 0; r1 < 4u; r1++)
        {
            if (r1 != a0 
             && UNK_020ECB54[r1][0] == UNK_020ECB54[a0][0]
             && UNK_02106FC8->unk94[r1] != NULL)
            {
                UNK_02106FC8->unk84[r1] = UNK_02106FC8->unk84[a0];
                break;
            }
        }
        if (r1 == 4u)
        {
            FreeToHeap(UNK_02106FC8->unk84[a0]);
            UNK_02106FC8->unk84[a0] = NULL;
        }
    }
    if (UNK_02106FC8->unk94[a0] != NULL)
    {
        FUN_020215C8(UNK_02106FC8->unk94[a0]);
        UNK_02106FC8->unk94[a0] = NULL;
    }
}
```

Reply:
wow thanks. It seems both using an `int` and turning `UNK_020ECB54` into a 2d array was necessary.